src/main/java/org/jitsi/videobridge/cc/AdaptiveTrackProjection.java

```diff
@@ -0,0 +1,307 @@
+     * The target quality index for this track projection.
+     */
+    private int targetIndex = RTPEncodingDesc.SUSPENDED_INDEX;
+
+    /**
+     * Ctor.
+     *
+     * @param source the {@link MediaStreamTrackDesc} that owns the packets
+     * that this instance filters.
+     */
+    AdaptiveTrackProjection(@NotNull MediaStreamTrackDesc source)
+    {
+        weakSource = new WeakReference<>(source);
+        targetSsrc = source.getRTPEncodings()[0].getPrimarySSRC();
+    }
+
+    /**
+     * @return the {@link MediaStreamTrackDesc} that owns the packets that this
+     * instance filters. Note that this may return null.
+     */
+    public MediaStreamTrackDesc getSource()
+    {
+        return weakSource.get();
+    }
+
+    /**
+     * @return the ideal quality for this track projection.
+     */
+    int getIdealIndex()
+    {
+        return idealIndex;
+    }
+
+    /**
+     * Update the ideal quality for this track projection.
+     *
+     * @param value the ideal quality for this track projection.
+     */
+    void setIdealIndex(int value)
+    {
+        idealIndex = value;
+    }
+
+    /**
+     * Gets the target index value for this track projection.
+     *
+     * @return the target index value for this track projection.
+     */
+    int getTargetIndex()
+    {
+        return targetIndex;
+    }
+
+    /**
+     * Sets the target index value for this track projection.
+     *
+     * @param value the new target index value for this track projection.
+     */
+    void setTargetIndex(int value)
+    {
+        targetIndex = value;
+    }
+
+    /**
+     * Determines whether an RTP packet needs to be accepted or not.
+     *
+     * @param rtpPacket the RTP packet to determine whether to accept or not.
+     * @return true if the packet is accepted, false otherwise.
+     */
+    public boolean accept(@NotNull RawPacket rtpPacket)
+    {
+        AdaptiveTrackProjectionContext
+            contextCopy = getContext(RawPacket.getPayloadType(rtpPacket));
+
+        int targetIndexCopy = targetIndex;
+        boolean accept = contextCopy.accept(rtpPacket, targetIndexCopy);
+
+        if (contextCopy.needsKeyframe() && targetIndexCopy > -1)
+        {
+            MediaStreamTrackDesc source = getSource();
+            if (source != null)
+            {
+                ((RTPTranslatorImpl) source
+                    .getMediaStreamTrackReceiver()
+                    .getStream()
+                    .getRTPTranslator())
+                    .getRtcpFeedbackMessageSender()
+                    .requestKeyframe(targetSsrc);
+            }
+        }
+
+        return accept;
+    }
+
+    /**
+     * Gets or creates the adaptive track projection context that corresponds to
+     * the payload type that is specified as a parameter. If the payload type
+     * is different from {@link #contextPayloadType}, then a new adaptive track
+     * projection context is created that is appropriate for the new payload
+     * type.
```

Comment:
so this will only track a context for one payload type at a time?  does it make sense to have this change/reset between payload types dynamically? for example, should we enforce only a single payload type per `AdaptiveTrackProjection`? Or have `AdaptiveTrackProjection` manage a context per payload type?

Reply:
The more general scheme and more correct approach would be to have the `AdaptiveTrackProjection` manage a context per payload type.. I wanted to implement the more general approach but we have no use case for that right now, and I don't have an easy way to test this dynamic payload type changes. So practically what happens is that we create the context once, and it remains the same throughout the life of the `AdaptiveTrackProjection` instance. It's not really possible to simplify and make the context final (=> create it during construction) because we need a payload type (could be VP9, could be H264, could be VP8) so it has to be created on packet receipt.

Reply:
gotcha, so the thinking is that: although we technically change this based on payload types, this is only because we need a payload type and don't get it until runtime and, practically, we only expect to get one per session so we don't really expect it to change. is that right?

(also, for questions like these, if you don't mind, let me resolve them--I ended up missing a bunch of these because they were collapsed in github and didn't see the answer because they were marked as resolved)